### ISSUE TITLE
no-undef: unexpectedly errors for valid 'with' block statements while no-with:0 rule is set

### ISSUE THREAD

**Issue by iskitz** [eslint/eslint#4978]
This issue's reproducible when using these, or similarly, valid `with` blocks
```javascript
with ({it:true}) {
    it;               // ESLint reports error 'it' is not defined no-undef
}

with (Math) {
    sqrt (4);         // ESLint reports error 'sqrt' is not defined no-undef
}
```

report | result
---------|----
_expected_ | no `no-undef` errors or warnings.
_received_ | `error 'it' is not defined no-undef`
_received_ | `error 'sqrt' is not defined no-undef`
    | .
__config__ | __.eslintrc__
[no-with](http://eslint.org/docs/rules/no-with) | 0
[no-undef](http://eslint.org/docs/rules/no-undef) | (default)

**Comment by eslintbot** [comment]
@iskitz Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by platinumazure** [comment]
I assume the same will happen when the object isn't literal:

```js
with (Math) {
    sqrt(4);
}
```

I'm not sure what we would be able to handle in solving this problem. Furthermore, I'm wondering if this really should belong to escope, since they should be able to create a new "scope" in the context of a with statement...

**Comment by iskitz** [comment]
@platinumazure 
> I assume the same will happen when the object isn't literal:

Yes. Same result.

**Comment by michaelficarra** [comment]
`no-undef` *should* warn here. Working as expected.

**Comment by platinumazure** [comment]
@michaelficarra Why do you say that? For the literal case, it should be
possible for static analysis to evaluate the object and add the keys to a
new scope. That said, for basically anything else I agree it would be
impossible.

**Comment by michaelficarra** [comment]
@platinumazure Only for certain keys. `with({[a]: 0});` completely ruins scoping as well.

**Comment by mysticatea** [comment]
Since ESLint is static analyzer, so we cannot analyze dynamic things correctly.

`WithStatement` is dynamic completely.

```js
Object.prototype[Symbol.unscopables] = {defined: true};
with ({defined: true}) {
    console.log(defined); // Uncaught ReferenceError: defined is not defined
}
```

**Comment by michaelficarra** [comment]
@mysticatea Ah that's right, I forgot unscopables could be changed at runtime.

**Comment by ilyavolodin** [comment]
I'm not sure about the discussion above, but as far as I can see, the rule is turned off, so it shouldn't warn. @iskitz Can you run eslint with `--debug` flag to see if you might have some other config enabling the rule back?

**Comment by platinumazure** [comment]
@ilyavolodin no-with is off but he is talking about no-undef, which is on and which he believes should not warn but does.

@iskitz Can you edit your original post to show no-undef in your configuration, as well as to more explicitly note what you expect and what you see? I think I have captured it correctly here but I need you to confirm or deny.

**Comment by iskitz** [comment]
@platinumazure
> Can you edit your original post to show no-undef in your configuration, as well as to more explicitly note what you expect and what you see?

Sure. Updated my original post and title to be more precise.

**Comment by iskitz** [comment]
Closing. No errors reported after updating to __ESLint v1.10.3__.

**Comment by michaelficarra** [comment]
That sounds like a bug.
